### DISCUSSION TITLE
pytest fails after FD seems installed

### DISCUSSION THREAD

**Discussion by obokhove** [firedrakeproject/firedrake#2652] [Firedrake support]
Firedrake seems installed (see log) but pytest fails; also running Junho's FD-program fails; any suggestions please; thank you (two logs attached):
onnobokhove@Os-MacBook-Pro werk % . /Users/onnobokhove/amtob/werk/firedrake/bin/activate
(firedrake) onnobokhove@Os-MacBook-Pro werk % cd $VIRTUAL_ENV/src/firedrake
(firedrake) onnobokhove@Os-MacBook-Pro firedrake % pytest tests/regression/ -k "poisson_strong or stokes_mini or dg_advection"
ImportError while loading conftest '/Users/onnobokhove/amtob/werk/firedrake/src/firedrake/tests/conftest.py'.
tests/conftest.py:9: in <module>
    from firedrake.utils import complex_mode
firedrake/__init__.py:80: in <module>
    from firedrake.output import *
firedrake/output.py:13: in <module>
    from .paraview_reordering import vtk_lagrange_tet_reorder,\
firedrake/paraview_reordering.py:5: in <module>
    from vtkmodules.vtkCommonDataModel import (
E   ModuleNotFoundError: No module named 'vtkmodules.vtkCommon
[configure.log](https://github.com/firedrakeproject/firedrake/files/10090389/configure.log)
DataModel'
(firedrake) onnobokhove@Os-MacBook-Pro firedrake % 
[firedrake-install.log](https://github.com/firedrakeproject/firedrake/files/10090386/firedrake-install.log)

**Comment by connorjward** [comment]
I think you have hit [this issue](https://github.com/firedrakeproject/firedrake/issues/2516). The known workaround is to run `pip uninstall vtk` and then `pip install vtk`.

**Comment by obokhove** [comment]
Did python3 -m pip uninstall vtk ; then ran: python3 pip install vtk ; reinstalled Firedrake.
Also tried it with python3.10 (no difference). What am I missing? Same error.


Still failing. (firedrake) onnobokhove@Os-MacBook-Pro firedrake % pytest tests/regression/ -k "poisson_strong or stokes_mini or dg_advection"
ImportError while loading conftest '/Users/onnobokhove/amtob/werk/firedrake/src/firedrake/firedrake/src/firedrake/tests/conftest.py'.
tests/conftest.py:9: in <module>
    from firedrake.utils import complex_mode
firedrake/__init__.py:80: in <module>
    from firedrake.output import *
firedrake/output.py:13: in <module>
    from .paraview_reordering import vtk_lagrange_tet_reorder,\
firedrake/paraview_reordering.py:5: in <module>
    from vtkmodules.vtkCommonDataModel import (
E   ModuleNotFoundError: No module named 'vtkmodules.vtkCommonDataModel'
(firedrake) onnobokhove@Os-MacBook-Pro firedrake %

**Reply by wence-** [reply]
You need to uninstall and install `vtk` _in_ the activated firedrake virtual environment (and then not reinstall firedrake). Whereas it seems like you did the vtk dance first, and then reinstalled firedrake (which will undo any changes that you might have made to the state of the vtk install).

**Comment by obokhove** [comment]
Thank you both. That works. Some experts (-) do not like homebrew, claiming it is insecure (they changed their mind with 1-2yrs) and insist FD is installed using singularity (dockers also not allowed due to security concerns). Could you comment, also on whether singularity is slower or leading to restrictions. I have FD now working on Mac OS with homebrew etc. Thank you.

**Reply by connorjward** [reply]
I can't comment on the security of homebrew, though I know a lot of people who use it without any problems. As for Singularity, it is supposedly quite difficult to get similar performance as you get with homebrew since you're dealing with an image/virtualisation sort of thing (not my speciality).

**Reply by obokhove** [reply]
Hi Connor and others: what is the performance reduction by using Singularity? Also when used in parallel on HPC?

**Reply by connorjward** [reply]
@JDBetteridge?

**Reply by JDBetteridge** [reply]
There is a guide to using Singularity/Apptainer using the Docker image as a starting point here: https://github.com/firedrakeproject/firedrake/wiki/singularity

It has been tested on ARCHER2, but due to various issues in scaling tests I don't have performance numbers. The biggest factor in performance will be getting the container to use the host MPI distribution. Feel free to contact me if you have questions about setting this up.
